compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEInsertDmaBdChain.cpp

```diff
@@ -0,0 +1,268 @@
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+using DmaChain = std::pair<AMDAIE::TileOp, AMDAIE::ConnectionOp>;
+
+/// Utility function to update `next_bd` and `start_bd` operands.
+LogicalResult updateChainOperands(
+    IRRewriter &rewriter, SmallVector<AMDAIE::NpuHalfDmaCpyNdOp> &dmaOps) {
+  // Nothing to do if the DMA chain length is one or less.
+  if (dmaOps.size() < 2) return success();
+
+  Value startBdId = dmaOps[0].getBdId();
+  Operation *parentOp = dmaOps[0]->getParentOp();
+  // Chain the DMA ops.
+  for (unsigned i = 0; i < dmaOps.size() - 1; ++i) {
+    AMDAIE::NpuHalfDmaCpyNdOp currDmaOp = dmaOps[i];
+    if (currDmaOp->getParentOp() != parentOp) {
+      return currDmaOp.emitError(
+          "DMA operations to be chained must belong to the same scope");
+    }
+    Value nextBdId = dmaOps[i + 1].getBdId();
+    // No token is produced at the beginning or middle of a chain.
+    TypeRange token = TypeRange{};
+    rewriter.setInsertionPointAfter(currDmaOp);
+    rewriter.create<AMDAIE::NpuHalfDmaCpyNdOp>(
+        currDmaOp.getLoc(), token, currDmaOp.getConnection(),
+        currDmaOp.getInput(), currDmaOp.getMixedOffsets(),
+        currDmaOp.getMixedSizes(), currDmaOp.getMixedStrides(),
+        currDmaOp.getBdId(), currDmaOp.getChannel(), nextBdId, startBdId);
+    for (auto &use : currDmaOp->getUses()) rewriter.eraseOp(use.getOwner());
+    rewriter.eraseOp(currDmaOp);
+  }
+  // Last DMA op in the chain.
+  AMDAIE::NpuHalfDmaCpyNdOp lastDmaOp = dmaOps.back();
+  if (lastDmaOp->getParentOp() != parentOp) {
+    return lastDmaOp.emitError(
+        "DMA operations to be chained must belong to the same scope");
+  }
+  Value nextBdId = nullptr;
+  rewriter.setInsertionPointAfter(lastDmaOp);
+  auto lastDmaOpChained = rewriter.create<AMDAIE::NpuHalfDmaCpyNdOp>(
+      lastDmaOp.getLoc(), lastDmaOp.getResultTypes(), lastDmaOp.getConnection(),
+      lastDmaOp.getInput(), lastDmaOp.getMixedOffsets(),
+      lastDmaOp.getMixedSizes(), lastDmaOp.getMixedStrides(),
+      lastDmaOp.getBdId(), lastDmaOp.getChannel(), nextBdId, startBdId);
+  rewriter.replaceOp(lastDmaOp, lastDmaOpChained.getResults());
+  return success();
+}
+
+/// Utility function to determine if chains can grow further
+/// or require breaking.
+///
+/// Example:
+/// - Chain X currently holds BD IDs: [4, 5, 6, 7]
+/// - Chain Y currently holds BD IDs: [0, 1, 2, 3]
+/// - A new BD ID (0) needs to be added to the front (due to reverse
+/// traversing) of chain X.
+///
+/// Conflict resolution:
+/// - Chain Y must be broken because BD ID 0 is already assigned to it
+/// and must be released.
+/// - Chain X is also broken to prevent the new added BD ID (0) from
+/// invalidating chain Y.
+///
+/// Result:
+/// - Break both chains X and Y.
+///   - Chain X: [0] (the newly added BD ID).
+///   - Chain Y: [] (emptied after breaking).
+void checkForChainsToBeBroken(
+    const uint32_t &currbdId, const DmaChain &currDmaChain,
```

Comment:
There's no need to pass basic types by reference:
```suggestion
    uint32_t currBdId, const DmaChain &currDmaChain,
```
Or, if you really want to ensure that `currbdId` isn't modified (locally):
```suggestion
    const uint32_t currBdId, const DmaChain &currDmaChain,
```
I do think most of the codebase (and upstream) use the former though.

And use `currBdId` instead of `currbdId`.